embedded-graphics/src/pixelcolor/raw/mod.rs

```diff
@@ -0,0 +1,225 @@
+//! impl PixelColor for RGBI {
+//!     type Raw = RawU4;
+//! }
+//!
+//! /// `From<RawU4>` is used by `Image` to construct RGBI colors.
+//! impl From<RawU4> for RGBI {
+//!     fn from(data: RawU4) -> Self {
+//!         Self(data)
+//!     }
+//! }
+//!
+//! /// Raw image data with 2 pixels per byte.
+//! const IMAGE_DATA: &[u8] = &[
+//!     0b0001_0010,
+//!     0b0100_1111,
+//! ];
+//!
+//! fn main() {
+//!     // Create new image with RGBI colors.
+//!     let image: Image<RGBI> = Image::new(IMAGE_DATA, 2, 2);
+//!
+//!     // In a real application the image could now be drawn to a display:
+//!     // display.draw(&image);
+//! #
+//! #   // Convert image to iterator of colors to make testing easier.
+//! #   let mut iter = image.into_iter().map(|Pixel(_, color)| color);
+//! #
+//! #   assert_eq!(iter.next(), Some(RGBI::new(false, false, true, false)));
+//! #   assert_eq!(iter.next(), Some(RGBI::new(false, true, false, false)));
+//! #   assert_eq!(iter.next(), Some(RGBI::new(true, false, false, false)));
+//! #   assert_eq!(iter.next(), Some(RGBI::new(true, true, true, true)));
+//! #   assert_eq!(iter.next(), None);
+//! }
+//! ```
+//!
+//! [`PixelColor`]: ../trait.PixelColor.html
+//! [`Raw`]: ../trait.PixelColor.html#associatedtype.Raw
+//! [`Image`]: ../../image/struct.Image.html
+
+mod iter;
+pub(crate) use iter::{RawDataIter, RawDataIterNext};
+
+/// Trait implemented by all `RawUx` types.
+pub trait RawData:
+    Sized
+    + private::Sealed
+    + RawDataIterNext<LittleEndian>
+    + RawDataIterNext<BigEndian>
+    + From<<Self as RawData>::Base>
+{
+    /// Integer base type.
+    ///
+    /// A primitive unsigned integer type that contains at least `BITS_PER_PIXEL`
+    /// bits.
+    type Base;
+
+    /// Bits per pixel.
+    const BITS_PER_PIXEL: usize;
+
+    /// Converts this raw data into the base type.
+    ///
+    /// If the primitive integer types used as the base type contains more bits
+    /// than used by this type the unused most significant bits are set to `0`.
+    fn into_inner(self) -> Self::Base;
+
+    /// Converts a `u32` into a `RawData` type.
+    ///
+    /// This method can be used to generically construct all `RawData` types from
+    /// the same integer type. If the width of the `RawData` type is less than
+    /// 32 bits only the least significant bits are used.
+    fn from_u32(value: u32) -> Self;
+}
+
+/// Dummy implementation for `()`.
+///
+/// `()` can be used as [`PixelColor::Raw`] if raw data conversion isn't required.
+///
+/// [`PixelColor::Raw`]: ../trait.PixelColor.html#associatedtype.Raw
+impl RawData for () {
+    type Base = ();
+
+    const BITS_PER_PIXEL: usize = 0;
+
+    fn into_inner(self) {}
+
+    fn from_u32(_value: u32) {}
+}
+
+impl private::Sealed for () {}
+
+macro_rules! impl_raw_data {
+    ($type:ident : $base_type:ident, $bpp:expr, $mask:expr, $bpp_str:expr) => {
+        #[doc = $bpp_str]
+        #[doc = "per pixel raw data."]
```

Comment:
Can you change this to describe the storage type used? Like

> 2 bits per pixel raw data, stored in a `u8`.

Reply:
> 2 bits per pixel raw data, stored in a `u8`.

Mentioning the storage type could be useful but I think this doesn't make it clear that it is only one 2bpp color value in an `u8` and not the packed 4 pixels per `u8` storage used in images.

Reply:
Can you expand the macro to allow a custom doc string to be passed in? I can't think of better names than `RawUx` _as long as_ it's made clear what the storage type behind each one is.